ompi/communicator/comm_cid.c

```diff
@@ -287,20 +301,146 @@ static volatile int64_t ompi_comm_cid_lowest_id = INT64_MAX;
 static int ompi_comm_cid_epoch = INT_MAX;
 #endif /* OPAL_ENABLE_FT_MPI */
 
+static int ompi_comm_ext_cid_new_block (ompi_communicator_t *newcomm, ompi_communicator_t *comm,
+                                        ompi_comm_extended_cid_block_t *new_block,
+                                        const void *arg0, const void *arg1, bool send_first, int mode,
+                                        ompi_request_t **req)
+{
+    pmix_info_t pinfo, *results = NULL;
+    size_t nresults;
+    opal_process_name_t *name_array;
+    char *tag = NULL;
+    size_t proc_count, cid_base = 0UL;
+    int rc, leader_rank;
+    pmix_proc_t *procs;
+
+    rc = ompi_group_to_proc_name_array (newcomm->c_local_group, &name_array, &proc_count);
+    if (OPAL_UNLIKELY(OMPI_SUCCESS != rc)) {
+        return rc;
+    }
+
+    switch (mode) {
+    case OMPI_COMM_CID_GROUP_NEW:
+        tag = (char *) arg0;
+        break;
+    case OMPI_COMM_CID_GROUP:
+        ompi_group_translate_ranks (newcomm->c_local_group, 1, &(int){0},
+                                    comm->c_local_group, &leader_rank);
+
+        tag = ompi_comm_extended_cid_get_unique_tag (&comm->c_contextidb, *((int *) arg0), leader_rank);
+        break;
+    case OMPI_COMM_CID_INTRA:
+        tag = ompi_comm_extended_cid_get_unique_tag (&comm->c_contextidb, -1, 0);
+        break;
+    }
+
+    PMIX_INFO_LOAD(&pinfo, PMIX_GROUP_ASSIGN_CONTEXT_ID, NULL, PMIX_BOOL);
```

Comment:
How scalable is the construction of the PMIx group ? What process is calling this function ?

Reply:
It uses the same PRRTE communication collective as the "modex" operation. All processes that are to be members of the group must call it. It basically is a data-less "modex" (aka barrier) that returns a unique context ID.

Reply:
In other words, slightly less scalable than MPI collectives on the local costs (because of the need to construct and destruct the group), then delegated to a different software that uses some communication capabilities, to finally be delegated to a RM with or without efficient support for collectives. Hopefully this CID selection is only rarely used and most calls are outside the critical path. Right ?